client/src/lib/user/state/migration/v3.ts

```diff
@@ -0,0 +1,133 @@
+import {omit, uniqBy} from 'ramda';
+import {UserEvent as CurrentUserEvent} from '../../../../../../shared/src/types/Event';
+import {LANGUAGE_TAG} from '../../../i18n';
+import {
+  PersistedState,
+  State as CurrentState,
+  UserState as CurrentUserState,
+} from '../state';
+
+// Types as they were in v3
+type V3PinnedSession = {
+  id: string;
+  expires: Date;
+};
+
+type V3CompletedSession = {
+  id: string;
+  hostId?: string;
+  exerciseId: string;
+  language: LANGUAGE_TAG;
+  type: V3SessionType;
+  mode: V3SessionMode;
+  hostProfile?: V3UserProfile;
+  completedAt: Date;
+};
+
+type V3UserProfile = {
+  displayName?: string;
+  photoURL?: string;
+};
+
+export enum V3SessionMode {
+  async = 'async',
+  live = 'live',
+}
+
+export enum V3SessionType {
+  private = 'private',
+  public = 'public',
+}
+
+export type V3PostPayload = {
+  sessionId: string;
+  exerciseId: string;
+  sharingId: string;
+  isPublic: boolean;
+  isAnonymous: boolean;
+  text: string;
+};
+
+type V3FeedbackPayload = {
+  like: boolean;
+  text?: string;
+};
+
+type V3PostEvent = {
+  timestamp: Date;
+  type: 'post';
+  payload: V3PostPayload;
+};
+
+type V3FeedbackEvent = {
+  timestamp: Date;
+  type: 'feedback';
+  payload: V3FeedbackPayload;
+};
+
+type V3UserEvent = V3PostEvent | V3FeedbackEvent;
+
+export type V3UserState = {
+  pinnedSessions?: Array<V3PinnedSession>;
+  completedSessions?: Array<V3CompletedSession>;
+  metricsUid?: string;
+  userEvents?: Array<V3UserEvent>;
+};
+
+export type V3State = {
+  userState: {[key: string]: V3UserState};
+};
+
+const migrateCompletedSessionsToEvents = (
+  completedSessions: V3CompletedSession[],
+): CurrentUserEvent[] => {
+  const sessions = uniqBy(s => s.id, completedSessions);
+  return sessions.map(s => ({
+    type: 'completedSession',
+    payload: omit(['completedAt'], s),
+    timestamp: new Date(s.completedAt),
+  })) as unknown as CurrentUserEvent[];
```

Comment:
Strange, the unknown thing again?
Is it because of ramda usage? cause I think I could do that without needing `as` at all, it was something like this:
```typescript
({ completedAt , ...s }) => ({
  type: 'completedSession',
  payload: s, 
  timestamp: new Date(completedAt)
})`

Reply:
Aah I get it, it is because of `type: V3SessionType;` it seems

Reply:
Does that kinda defeats the purpose of using the version types?

Reply:
This is a special case with the unions, same issue in the user state file where I added a `getTypedEvent` to cast

Reply:
A way to make the `as` unnecessary is to update the `V3CompletedSession` fields to:

```typescript
type: SessionType.private | SessionType.public;
mode: SessionMode.async | SessionMode.live;
```

Reply:
Hmmm... not sure if I understand, on the `getTypedEvent` is different no?
Like, you're not returning `as unkown as ...` there, the type check actually occurs and still works...